File: php/src/Advertisements/Advertisement/Application/Command/DisableAdvertisement/DisableAdvertisementUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\Advertisement\Application\Command\DisableAdvertisement;

use Demo\App\Advertisements\Advertisement\Domain\AdvertisementRepository;
use Demo\App\Advertisements\Advertisement\Domain\Services\AdvertisementSecurityService;
use Demo\App\Advertisements\Advertisement\Domain\ValueObjects\AdvertisementId;
use Demo\App\Advertisements\Shared\ValueObjects\UserId;
use Demo\App\Framework\Database\TransactionManager;
use Exception;

final class DisableAdvertisementUseCase
{
    public function __construct(
        private AdvertisementRepository      $advertisementRepository,
        private AdvertisementSecurityService $securityService,
        private TransactionManager           $transactionManager,
    ) {}

    /**
     * @throws Exception
     */
    public function execute(DisableAdvertisementCommand $command): void
    {
        $this->transactionManager->beginTransaction();

        try{
            $advertisement = $this->advertisementRepository->findByIdOrFail(new AdvertisementId($command->advertisementId));

            if ('admin' === $command->securityUserRole) {
                $this->securityService->verifyAdminUserCanManageAdvertisement(
                    new UserId($command->securityUserId),
                    $advertisement,
                );
            }

            if ('user' === $command->securityUserRole) {
                $this->securityService->verifyMemberUserCanManageAdvertisement(
                    new UserId($command->securityUserId),
                    $advertisement,
                );
            }

            $advertisement->disable();

            $this->advertisementRepository->save($advertisement);
        } catch (Exception $exception) {
            $this->transactionManager->rollback();
            throw $exception;
        }
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/application/command/disableadvertisement/DisableAdvertisementUseCase.kt
package advertisements.advertisement.application.command.disableadvertisement

import advertisements.advertisement.domain.exceptions.AdvertisementNotFoundException
import advertisements.advertisement.domain.AdvertisementRepository
import advertisements.advertisement.domain.services.AdvertisementSecurityService
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.shared.value_object.UserId

class DisableAdvertisementUseCase(
    private val advertisementRepository: AdvertisementRepository,
    private val advertisementSecurityService: AdvertisementSecurityService
) {
    fun execute(disableAdvertisementCommand: DisableAdvertisementCommand) {
        val advertisementId = AdvertisementId(disableAdvertisementCommand.advertisementId)
        val advertisement = advertisementRepository.findById(advertisementId)

        if (null === advertisement) {
            throw AdvertisementNotFoundException.withId(advertisementId.value())
        }

        advertisementSecurityService.verifyAdminUserCanManageAdvertisement(
            UserId(disableAdvertisementCommand.securityUserId),
            advertisement
        )

        advertisement.disable()

        advertisementRepository.save(advertisement)
    }
}
